File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    

public boolean isValidBST(TreeNode root) {
    return bfs(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
}

private boolean bfs(TreeNode currentNode, int leftBoundary, int rightBoundary) {
    if (currentNode == null)
        return true;

    if(currentNode.val < rightBoundary && currentNode.val > leftBoundary){
        boolean rightChild = bfs(currentNode.right, currentNode.val, rightBoundary);
        boolean leftChild = bfs(currentNode.left, leftBoundary, currentNode.val);
        return rightChild && leftChild;

    }else{
        return false;
    }

}
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def __init__(self):
        self.lca = -505

    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        twoNodesSet = {p.val, q.val}
        self.solve(root, twoNodesSet)
        return self.lca

    def solve(self, node, twoNodesSet):
        if not node:
            return -505   
        a = self.solve(node.left, twoNodesSet)
        b = self.solve(node.right, twoNodesSet)
        firstSet = {a,b}
        secondSet = {a,node.val}
        thirdSet = {b,node.val}
        if firstSet == twoNodesSet or secondSet == twoNodesSet or thirdSet == twoNodesSet:
            self.lca = node
        if a in twoNodesSet:
            return a
        elif b in twoNodesSet:
            return b
        elif node.val in twoNodesSet:
            return node.val
        else:
            return -505            

            



        